core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -99,21 +139,51 @@ public void registerCacheEventListener(CacheEventListener<? super K, ? super V>
    */
   @Override
   public void deregisterCacheEventListener(CacheEventListener<? super K, ? super V> listener) {
-    if (!registeredListeners.remove(new EventListenerWrapper(listener,
-        EventFiring.ASYNCHRONOUS, EventOrdering.UNORDERED, EnumSet.allOf(EventType.class)))) {
-      throw new IllegalStateException("Unknown cache event listener: " + listener);
+    EventListenerWrapper wrapper = new EventListenerWrapper(listener,
+        EventFiring.ASYNCHRONOUS, EventOrdering.UNORDERED, EnumSet.allOf(EventType.class));
+    boolean removed = false;
+    for (EventListenerWrapper listenerWrapper : aSyncListenersSet) {
+      if(listenerWrapper.equals(wrapper)) {
+        aSyncListenersSet.remove(listenerWrapper);
+        removed = true;
+        if(listenerWrapper.config.orderingMode() == EventOrdering.ORDERED) {
+          orderedListenerCount.decrementAndGet();
+        }
+        break;
+      }
     }
-    if (!hasListeners()) {
+    if (!removed) {
+      for (EventListenerWrapper listenerWrapper : syncListenersSet) {
+        if(listenerWrapper.equals(wrapper)) {
+          syncListenersSet.remove(listenerWrapper);
+          removed = true;
+          if(listenerWrapper.config.orderingMode() == EventOrdering.ORDERED) {
+            orderedListenerCount.decrementAndGet();
+          }
+          break;
+        }
+      }
+    }
+    synchronized (this) {
+      if (!hasListeners()) {
+        if (eventThreadLocal instanceof EventThreadLocalImpl) {
+          eventThreadLocal = new NoOpEventThreadLocalImpl();
+        }
+      }
       store.disableStoreEventNotifications();
     }
+    if(!removed) {
+      throw new IllegalStateException("Unknown cache event listener: " + listener);
+    }
   }
 
   // TODO this should be really the shutdown method for the service
   @Override
   public void releaseAllListeners() {
-    for (EventListenerWrapper wrapper: registeredListeners) {
-      registeredListeners.remove(wrapper);
-    }
+    orderedListenerCount.set(0);
+    aSyncListenersSet.clear();
+    syncListenersSet.clear();
+    eventThreadLocal.cleanUp();
```

Comment:
Not what I meant with the comment on the line above. After this call, there are no listeners left so ... ??